### PR TITLE
[inductor] Add structural divisibility analysis to statically_known_multiple_of 

### LINKED ISSUE #177146: Divisibility constraints are not propagated through symbolic expressions in Inductor
## Summary

`torch._check(x.shape[-1] % 16 == 0)` improves `torch.compile(dyanmic=True)` performance for RMS Norm, see #175755.
However, the divisibiliy is not propagated  to other symblic expres such as `xnumel = shape[0] * shape[-1]` in pointwise kernels. 
This causes significant performance loss when normalization kernel is segmented into `reduction + pointwise`.

## Root Cause

`statically_known_multiple_of(s27*s77, 16)` in `torch/_inductor/sizevars.py` asks sympy to evaluate `Eq(Mod(s27*s77, 16), 0)`. Even though the ShapeEnv has the axiom `Eq(Mod(s27, 16), 0)` (from the `torch._check`), sympy cannot deduce that a product `s27*s77` is divisible by 16 when one of its factors `s27` is — it does not perform the algebraic step: **if `a % n == 0` then `(a*b) % n == 0`**.

## Repro (on GB200)

```python
import torch

@torch.compile(dynamic=True, backend="inductor")
def f(x, w):
    torch._check(x.shape[-1] % 16 == 0)  # hidden_dim is always aligned
    rms = torch.sqrt(torch.mean(x.float() ** 2, dim=-1, keepdim=True) + 1e-6)
    return torch.nn.functional.gelu((x.float() / rms) * w.float()).half()

x = torch.randn([127, 262144], device="cuda", dtype=torch.float16)
w = torch.randn([262144], device="cuda", dtype=torch.float16)
f(x, w)
```

The graph segments into:
1. **Reduction kernels** (`2 kernels`): computes mean along `dim=-1` — `rnumel = shape[-1]` gets `tt.divisibility=16` correctly
2. **Pointwise kernel** (`1 kernel`): normalize + scale + GELU — `xnumel = shape[0] * shape[-1]` does **not** get `tt.divisibility=16`

The pointwise kernel's `xnumel` argument missing `tt.divisibility` prevents Triton from optimizing loads/stores, causing a **1.66x slowdown** on the pointwise kernel.

## Proposed Fix (Immediate)

In `statically_known_multiple_of`, when the direct sympy check fails and `numerator` is
a `sympy.Mul`, decompose it into factors and check if any individual factor is known to
be divisible by the denominator.

```diff
--- a/torch/_inductor/sizevars.py
+++ b/torch/_inductor/sizevars.py
@@ -417,7 +417,22 @@ class SizeVarAllocator:

         expr = sympy.Eq(Mod(numerator, denominator), 0)
-        return self.statically_known_true(expr)
+        if self.statically_known_true(expr):
+            return True
+
+        # If numerator is a product (a * b * ...), check if any factor is a
+        # multiple of the denominator. Handles e.g. xnumel = shape[0] * shape[-1]
+        # when torch._check(x.shape[-1] % 16 == 0) is asserted.
+        if isinstance(numerator, sympy.Mul) and isinstance(denominator, (int, sympy.Integer)):
+            denom_int = int(denominator)
+            for factor in numerator.args:
+                if isinstance(factor, (int, sympy.Integer)):
+                    if int(factor) % denom_int == 0:
+                        return True
+                else:
+                    factor_expr = sympy.Eq(Mod(factor, denominator), 0)
+                    if self.statically_known_true(factor_expr):
+                        return True
+
+        return False
```

## Benchmark Results (GB200, SM 10.0)

`[127, 262144]` fp16, RMSNorm + GELU with `torch._check(x.shape[-1] % 16 == 0)`:

| Kernel | Before fix | After fix | Speedup |
|--------|-----------|-----------|---------|
| reduction (triton_red) | 39.4 us | 34.3 us | 1.15x |
| persistent (triton_per) | 1.9 us | 1.4 us | 1.36x |
| **pointwise (triton_poi)** | **108.7 us** | **65.3 us** | **1.66x** |
| **Total** | **150.0 us** | **101.1 us** | **1.48x** |

### Error logs

_No response_

### Versions

PyTorch version: 2.12.0a0+git5710dd2
Is debug build: False
CUDA used to build PyTorch: 13.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (aarch64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Jan 22 2026, 20:57:42) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-1008-nvidia-64k-aarch64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 13.2.46

cc @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by liqiangxl**
Add `_is_multiple_of()` that recurses over sympy expression structure (Mul, Add, FloorDiv, Mod) to prove divisibility, before falling back to statically_known_true. This is a demand-driven variant of the modular arithmetic analysis used in Halide and TVM.

Unlocks tt.divisibility hints for product expressions like `shape[0] * shape[1]` when `torch._check(shape[1] % 16 == 0)`.

Fixes https://github.com/pytorch/pytorch/issues/177146

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/sizevars.py [hunk 1]
```diff
@@ -404,6 +404,54 @@ def statically_known_gt(self, left: Expr, right: Expr | int) -> bool:
...
+        if isinstance(numerator, sympy.Add):
+            if all(self._is_multiple_of(term, denominator) for term in numerator.args):
+                return True
+
+        # Rule 4 — FloorDiv(a, b): if a is multiple of b*n
+        if isinstance(numerator, FloorDiv):
+            a, b = numerator.args
+            if isinstance(b, (int, sympy.Integer)):
+                if self._is_multiple_of(a, int(b) * denominator):
+                    return True
```
Comment (jansel) [inline torch/_inductor/sizevars.py:440]: Could you simplify to:
```
return sympy.gcd(numerator, denominator) == denominator
```

**Review by jansel — CHANGES_REQUESTED** [review]
Fix lints

File: torch/_inductor/sizevars.py (hunk 1 above)
Comment (liqiangxl) [reply torch/_inductor/sizevars.py:440]: Thanks for the suggestion. It doesn't work with a simple repalcement using `sympy.gcd`.

**Why doesn't work:**
The divisibility facts are stored as axioms in ShapeEnv, `shape_env.axioms[Eq(Mod(s1, 16), 0)] = True`. 
This does not change s1's sympy representation, it is still a bare symbol.
So `sympy.gcd(s1 * s2, 16)` returns 1 instead of 16, even though s1 is axiomatically a multiple of 16.

**To make it work:** we need the following steps which seems non-trivial.
1. Scan ShapeEnv axioms for all `Mod(symbol, k) == 0` facts,
2. Substitute each such symbol with `k * fresh_symbol`,
3. Call `sympy.gcd` on the rewritten expression.

File: torch/_inductor/sizevars.py [hunk 2]
```diff
@@ -404,6 +404,54 @@ def statically_known_gt(self, left: Expr, right: Expr | int) -> bool:
...
 
+    def _is_multiple_of(self, numerator: Expr, denominator: int) -> bool:
+        """
+        Structural divisibility check: returns True only if numerator is
+        provably a multiple of denominator.  Recurses over sympy expression
+        structure before falling back to statically_known_true.
+        """
+        # Rule 1 — concrete value
+        if isinstance(numerator, (int, sympy.Integer)):
+            return int(numerator) % denominator == 0
```
Comment (laithsakka) [inline torch/_inductor/sizevars.py:415]: 1. do we want to use  class Mod(sympy.Function).

is this not handled by statically_known_true trivially? or we are just trying to be faster.

File: torch/_inductor/sizevars.py [hunk 3]
```diff
@@ -404,6 +404,54 @@ def statically_known_gt(self, left: Expr, right: Expr | int) -> bool:
...
+        if isinstance(numerator, sympy.Mul):
+            for factor in numerator.args:
+                if self._is_multiple_of(factor, denominator):
+                    return True
+            # Also check if combined constant factors are divisible
+            const = 1
+            for factor in numerator.args:
+                if isinstance(factor, (int, sympy.Integer)):
+                    const *= int(factor)
+            if const != 1 and const % denominator == 0:
```
Comment (laithsakka) [inline torch/_inductor/sizevars.py:427]: why  const != 1 ?

File: torch/_inductor/sizevars.py [hunk 4]
```diff
@@ -416,6 +464,10 @@ def statically_known_multiple_of(
         if len(free_symbols(numerator)) > 20:
             return False
 
+        if isinstance(denominator, (int, sympy.Integer)):
```
Comment (laithsakka) [inline torch/_inductor/sizevars.py:467]: mm which one shall go first this or self.statically_known_true(expr)...
the cheaper probably i cant tel which one is cheaper, but given that self.statically_known_true is not giving regressions shall we do it first?

File: torch/_inductor/sizevars.py [hunk 5]
```diff
@@ -404,6 +404,54 @@ def statically_known_gt(self, left: Expr, right: Expr | int) -> bool:
         expr = left > right
         return self.statically_known_true(expr)
 
+    def _is_multiple_of(self, numerator: Expr, denominator: int) -> bool:
```
Comment (laithsakka) [inline torch/_inductor/sizevars.py:407]: FYI many of the ideas here work also if denominator is not int, consider generalizing but its fine as is we can do that incrementally as we need.

**Review by laithsakka — APPROVED** [review]

**Review by jansel — APPROVED** [review]

File: torch/_inductor/sizevars.py (hunk 2 above)
Comment (liqiangxl) [reply torch/_inductor/sizevars.py:415]: 1. No, it goes to `expr = sympy.Eq(Mod(numerator, denominator), 0)` even both `numerator` and `denominator` are `int`.
2. It is a performance optimization, but also required to terminate the recursion, e.g., `(ks * 1024) % 16` will iterate to `ks % 16` and `1024 % 16`

File: torch/_inductor/sizevars.py (hunk 3 above)
Comment (liqiangxl) [reply torch/_inductor/sizevars.py:427]: It was initilized to 1, stays at 1 means no const factors.
Not absolutely required. Added to skip evaluating `1 % denominator`.

File: torch/_inductor/sizevars.py (hunk 4 above)
Comment (liqiangxl) [reply torch/_inductor/sizevars.py:467]: good point. I don't have data about the cost, but the following code comment seems to suggest that`statically_known_true ` is not cheap.
```
        # The reason we skip compute here is to avoid the cost of trying to eval this symbolically.
        # see https://github.com/sympy/sympy/issues/28200

**Comment by jansel** [comment]
@claude please diagnose the CI failures

File: torch/_inductor/sizevars.py (hunk 4 above)
Comment (laithsakka) [reply torch/_inductor/sizevars.py:467]: right but _is_multiple_of calls it many times vs one time in the other branch 
(each leaf in the recursion is         return self.statically_known_true(expr))
